Fixed improper rendering of ThumbimageSource in Slider

NirmalKumarYuvaraj opened the pull request:
### Issue Details

The ThumbImage set on the slider is rendering too large in view.

### Root Cause

The image source dimensions are not being properly scaled .

### Description of Change

Implementing a custom resizing logic to resize the image before rendering it on the slider to ensure optimal display and improve user experience.

Validated the behaviour in the following platforms

- [x] Android
- [ ] Windows
- [x] iOS
- [x] Mac

### Issues Fixed

Fixes #13258

### Output

| Before | After |
| ------ | ----- |
|<Image src="https://github.com/user-attachments/assets/5d785538-98bd-4691-b2fe-809c1994d754">|<image src="https://github.com/user-attachments/assets/5b08f015-3a2a-414c-a618-f0dd00e6946d">|

NirmalKumarYuvaraj committed:
Fixed slider thumb image scaling issue

NirmalKumarYuvaraj committed:
Modified the code changes

NirmalKumarYuvaraj committed:
Updated Images

NirmalKumarYuvaraj committed:
updated code changes

NirmalKumarYuvaraj committed:
Modified resizing logic

NirmalKumarYuvaraj committed:
Updated code changes

NirmalKumarYuvaraj committed:
Updated code changes

jsuarezruiz reviewed (changes requested):
Seeing the behavior on Android and iOS makes me doubt, does it not happen on Windows? Is it not necessary to resize?

NirmalKumarYuvaraj commented on src/Core/src/Platform/iOS/SliderExtensions.cs:
```diff
@@ -44,14 +46,14 @@ public static void UpdateThumbColor(this UISlider uiSlider, ISlider slider)
-				var thumbImage = result?.Value;
-
+				var thumbImageSize = result?.Value.Size ?? CGSize.Empty;
+				const float TARGET_SIZE = 28f;
```
I have updated the measure logic. Please let me know if you have any concerns.

NirmalKumarYuvaraj committed:
updated code changes

jsuarezruiz commented:
/azp run

NirmalKumarYuvaraj commented on src/Core/src/Platform/Android/SliderExtensions.cs:
```diff
@@ -47,21 +50,35 @@ public static void UpdateThumbColor(this SeekBar seekBar, ISlider slider) =>
+					if (thumbDrawable is BitmapDrawable bitmapDrawable && bitmapDrawable.Bitmap is { } bitmap)
+					{
+						// Define the target size for the thumb image
+						const int TARGET_SIZE = 48; // 48dp - default size of the thumb in Android
```
@jsuarezruiz ,Using the default thumbsize causes the image to be appear bigger in the view. So i have used the value 48 which will be the default height and width of the thumb.Please let me know if you have any concerns.

<image src="https://github.com/user-attachments/assets/adae26cf-8348-4c96-b96f-5dbb3ac790d3" height=600 width=300>

code snippet:
```C#
var defaultThumbSize = context.GetDrawable(Resource.Drawable.abc_seekbar_thumb_material);
var thumbImage = bitmap.Downsize(defaultThumbSize!.IntrinsicWidth, defaultThumbSize.IntrinsicHeight);
```

mattleibow replied:
Can the const be at the top of the file along with a comment as to where the number came from? Is it from some android xml theme file?

mattleibow replied:
Also, what do you mean by "Using the default thumbsize causes the image to be appear bigger in the view"?

I think Javier is meaning can you read the value from the slider? Like, is there a slider.GetThumb method?

NirmalKumarYuvaraj replied:
@mattleibow , The seekBar.Thumb?.IntrinsicHeight and IntrinsicWidth return a value of 47 before resizing the thumb. To avoid pixel discrepancies, I rounded this to 48 and made it a constant value. I researched documentation regarding the thumb size. Material Design 2 (circular) suggests that the radius should be 10dp, whereas Material Design 3 (vertically positioned rectangle) recommends a height of 44dp and a width of 4dp. Could you please confirm which approach we should proceed with?

Material 2 - https://m2.material.io/components/sliders/android#discrete-slider
Material 3 - https://m3.material.io/components/sliders/specs

jsuarezruiz commented on src/Controls/tests/TestCases.iOS.Tests/snapshots/ios/SliderShouldChangeThumbImageAndResetIt.png:
Test failing on iOS with small differences:
![image](https://github.com/user-attachments/assets/61c39bee-7899-4a8a-9663-331cb2590a44)

NirmalKumarYuvaraj replied:
@jsuarezruiz , i have attached the snapshots. please let me know if you have any concerns.

NirmalKumarYuvaraj committed:
Updated test case images

jsuarezruiz commented:
/azp run

jsuarezruiz commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue13258.cs:
```diff
@@ -0,0 +1,25 @@
+
+		[Test]
+		[Category(UITestCategories.Slider)]
+		public void SliderThumbImageShouldBeScaled()
```
Pending snapshot on Android:
![image](https://github.com/user-attachments/assets/cc6ce178-0c44-4db4-99dc-e80dbdcfc1d9)
Could you commit it?

jsuarezruiz replied:
The thumb image is clearly blurry. From a big image, related with the resize code?

NirmalKumarYuvaraj replied:
@jsuarezruiz , I have updated the size of the thumb to align with the Material 2 design specifications. I implemented new logic to preserve the quality of the image, as it was previously blurry due to resizing from a large image. please let me know if you have any concerns.

jfversluis commented:
@NirmalKumarYuvaraj please address feedback, thanks!

NirmalKumarYuvaraj committed:
Updated code changes

NirmalKumarYuvaraj committed:
Added snap shots

NirmalKumarYuvaraj commented:
> @NirmalKumarYuvaraj please address feedback, thanks!

@jfversluis , I have addressed the feedback and modified the changes. Please let me know if you have any concern.

jsuarezruiz commented on src/Core/src/Platform/Android/SliderExtensions.cs:
```diff
@@ -47,37 +50,62 @@ public static void UpdateThumbColor(this SeekBar seekBar, ISlider slider) =>
+			thumbDrawable.SetBounds(0, 0, thumbSize, thumbSize);
+			thumbDrawable.Draw(canvas);
+
+			BitmapDrawable finalDrawable = new BitmapDrawable(context.Resources, bitmap);
```
Used using statements to ensure proper disposal of BitmapDrawable:
```
using (BitmapDrawable finalDrawable = new BitmapDrawable(context.Resources, bitmap))
{
    seekBar.SetThumb(finalDrawable);
}
```

NirmalKumarYuvaraj replied:
@jsuarezruiz ,  I have updated the code changes. please let me know if you have any concerns.

jsuarezruiz commented on src/Core/src/Platform/Android/SliderExtensions.cs:
```diff
@@ -47,37 +50,62 @@ public static void UpdateThumbColor(this SeekBar seekBar, ISlider slider) =>
+			int thumbSize = (int)context.ToPixels(TARGET_SIZE);
+
+			Bitmap bitmap = Bitmap.CreateBitmap(thumbSize, thumbSize, Bitmap.Config.Argb8888!);
+			Canvas canvas = new Canvas(bitmap);
```
Can also use a using statement with the Bitmap and the Canvas.

NirmalKumarYuvaraj replied:
@jsuarezruiz , I have updated the code changes. please let me know if you have any concerns.

jsuarezruiz commented on src/Graphics/src/Graphics/Platforms/iOS/UIImageExtensions.cs:
```diff
@@ -59,5 +60,23 @@ public static UIImage NormalizeOrientation(this UIImage target, bool disposeOrig
+			if (maxResizeFactor > 1 && !shouldScaleUp)
+				return sourceImage;
+
+			return UIImage.FromImage(sourceImage.CGImage, sourceImage.CurrentScale / maxResizeFactor, sourceImage.Orientation);
```
To avoid create a new UIImage instance, could we use `UIGraphics.BeginImageContext `to create a context for drawing the scaled image, resize it and use `UIGraphics.GetImageFromCurrentImageContext` to retrieve the resized image?

NirmalKumarYuvaraj replied:
@jsuarezruiz , I previously used this approach to resize the image. However, regarding other similar resizing issues in iOS, Mattie commented that this method may cause memory issues and suggested using the current approach I implemented for resizing. I have shared the comment link below. Please let me know if you have any concerns.

Link - https://github.com/dotnet/maui/pull/26016/files/118ad1329919d37702ffb42dc08a59edd5d9dabe#r1871770914

NirmalKumarYuvaraj committed:
Update SliderExtensions.cs

jsuarezruiz commented:
/rebase

jsuarezruiz commented:
/azp run

rmarinho commented:
/rebase

rmarinho commented on src/Core/src/Platform/Android/SliderExtensions.cs:
```diff
@@ -47,37 +50,66 @@ public static void UpdateThumbColor(this SeekBar seekBar, ISlider slider) =>
 			var context = seekBar.Context;
-
-			if (context == null)
+			if (context is null || !seekBar.IsAlive())
 				return;
```
```suggestion
			if (context is null || !seekBar.IsAlive())
			{
				return;
			}
```

NirmalKumarYuvaraj replied:
I have updated the code changes

rmarinho commented on src/Core/src/Platform/Android/SliderExtensions.cs:
```diff
@@ -47,37 +50,66 @@ public static void UpdateThumbColor(this SeekBar seekBar, ISlider slider) =>
+		{
+			seekBar.SetThumb(context.GetDrawable(Resource.Drawable.abc_seekbar_thumb_material));
+
+			if (slider.ThumbColor == null && context.Theme != null)
```
```suggestion
			if (slider.ThumbColor is null && context.Theme is not null)
```

NirmalKumarYuvaraj replied:
Addressed the changes requested

rmarinho commented on src/Graphics/src/Graphics/Platforms/iOS/UIImageExtensions.cs:
```diff
@@ -59,5 +60,23 @@ public static UIImage NormalizeOrientation(this UIImage target, bool disposeOrig
+
+		internal static UIImage ResizeImageSource(this UIImage sourceImage, nfloat maxWidth, nfloat maxHeight, CGSize originalImageSize, bool shouldScaleUp = false)
+		{
+			if (sourceImage is null || sourceImage.CGImage is null)
+				return null;
```
```suggestion
			if (sourceImage is null || sourceImage.CGImage is null)
			{
				return null;
			}
```

NirmalKumarYuvaraj replied:
Addressed the changes requested

rmarinho commented on src/Graphics/src/Graphics/Platforms/iOS/UIImageExtensions.cs:
```diff
@@ -59,5 +60,23 @@ public static UIImage NormalizeOrientation(this UIImage target, bool disposeOrig
+
+			float maxResizeFactor = (float)Math.Min(maxWidth / sourceSize.Width, maxHeight / sourceSize.Height);
+
+			if (maxResizeFactor > 1 && !shouldScaleUp)
+				return sourceImage;
```
```suggestion
			if (maxResizeFactor > 1 && !shouldScaleUp)
			{
				return sourceImage;
			}
```

NirmalKumarYuvaraj replied:
Addressed the changes requested

rmarinho commented on src/Core/src/Platform/iOS/SliderExtensions.cs:
```diff
@@ -44,14 +46,14 @@ public static void UpdateThumbColor(this UISlider uiSlider, ISlider slider)
 				var result = await service.GetImageAsync(thumbImageSource, scale);
-				var thumbImage = result?.Value;
-
+				var thumbImageSize = result?.Value.Size ?? CGSize.Empty;
```
Does it make sense if the ThumbImageSize is Empty to still try to resize the image, or in that case we should go to the case where we use `var thumbImage = result?.Value;`

rmarinho reviewed (changes requested):
Small comments and improvements

NirmalKumarYuvaraj committed:
Updated code changes

jsuarezruiz commented:
/azp run

jsuarezruiz reviewed (changes requested):
There are failing Slider tests:
<img width="349" alt="image" src="https://github.com/user-attachments/assets/bb676910-3d30-43ab-ba4e-fc89b6859052" />

Example:
![image](https://github.com/user-attachments/assets/d1413ffb-bb3d-4e8b-9a18-09a751370e99)

Could you review it?

NirmalKumarYuvaraj committed:
updated snaps

NirmalKumarYuvaraj committed:
added issue link to script file

NirmalKumarYuvaraj commented on src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue13258.cs:
```diff
@@ -0,0 +1,25 @@
+﻿#if TEST_FAILS_ON_WINDOWS
```
@MartyIX , Added windows issue link

jsuarezruiz commented:
/rebase

jsuarezruiz commented:
/azp run

NirmalKumarYuvaraj committed:
Resaved Mac snaps

jsuarezruiz commented:
/rebase

jsuarezruiz commented:
/azp run MAUI-UITests-public

mattleibow reviewed (changes requested):
Mostly good, just a nit on the comment and const location.

And then we need a IsAlive check after the await returns.

NirmalKumarYuvaraj committed:
Update SliderExtensions.cs

jsuarezruiz commented:
/azp run MAUI-UITests-public

NirmalKumarYuvaraj committed:
Update SliderExtensions.cs

jsuarezruiz reviewed (commented):
@NirmalKumarYuvaraj Could you rebase and fix the conflicts? Thanks in advance.
